drivers/rtc/rtc_rv3028.c

```diff
@@ -0,0 +1,902 @@
+			return -ENODEV;
+		}
+
+		err = gpio_pin_configure_dt(&config->gpio_int, GPIO_INPUT);
+		if (err) {
+			LOG_ERR("failed to configure GPIO (err %d)", err);
+			return -ENODEV;
+		}
+
+		gpio_init_callback(&data->int_callback, rv3028_int_handler,
+				   BIT(config->gpio_int.pin));
+
+		err = gpio_add_callback_dt(&config->gpio_int, &data->int_callback);
+		if (err) {
+			LOG_ERR("failed to add GPIO callback (err %d)", err);
+			return -ENODEV;
+		}
+
+		data->dev = dev;
+		data->work.handler = rv3028_work_cb;
+	}
+#endif /* RV3028_INT_GPIOS_IN_USE */
+
+	err = rv3028_read_reg8(dev, RV3028_REG_STATUS, &val);
+	if (err) {
+		return -ENODEV;
+	}
+
+	if (val & RV3028_STATUS_AF) {
+		LOG_WRN("an alarm may have been missed");
+	}
+
+	/* Refresh the settings in the RAM with the settings from the EEPROM */
+	err = rv3028_enter_eerd(dev);
+	if (err) {
+		return -ENODEV;
+	}
+	err = rv3028_refresh(dev);
+	if (err) {
+		return -ENODEV;
+	}
+
+	/* Configure the CLKOUT register */
+	val = FIELD_PREP(RV3028_CLKOUT_FD, config->cof) |
+	      (config->cof != RV3028_CLKOUT_FD_LOW ? RV3028_CLKOUT_CLKOE : 0);
+	err = rv3028_update_cfg(dev,
+				RV3028_REG_CLKOUT,
+				RV3028_CLKOUT_FD | RV3028_CLKOUT_CLKOE,
+				val);
+	if (err) {
+		return -ENODEV;
+	}
+
+	err = rv3028_update_cfg(dev,
+				RV3028_REG_BACKUP,
+				RV3028_BACKUP_TCE | RV3028_BACKUP_TCR | RV3028_BACKUP_BSM,
+				config->backup);
+	if (err) {
+		return -ENODEV;
+	}
+
+	err = rv3028_update_reg8(dev, RV3028_REG_CONTROL1, RV3028_CONTROL1_WADA,
+				 RV3028_CONTROL1_WADA);
+	if (err) {
+		return -ENODEV;
+	}
+
+	/* Disable the alarms */
+	err = rv3028_update_reg8(dev,
+				 RV3028_REG_CONTROL2,
+				 RV3028_CONTROL2_AIE | RV3028_CONTROL2_UIE,
+				 0);
+	if (err) {
+		return -ENODEV;
+	}
+
+	err = rv3028_read_regs(dev, RV3028_REG_ALARM_MINUTES, regs, sizeof(regs));
+	if (err) {
+		return -ENODEV;
+	}
+
+	regs[0] |= RV3028_ALARM_MINUTES_AE_M;
+	regs[1] |= RV3028_ALARM_HOURS_AE_H;
+	regs[2] |= RV3028_ALARM_DATE_AE_WD;
+
+	err = rv3028_write_regs(dev, RV3028_REG_ALARM_MINUTES, regs, sizeof(regs));
+	if (err) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_RTC_RV3028_EEPROM
+#define RV3028_EEPROM_INIT(node, rtc_inst)                                                         \
+	static const struct rv3028_eeprom_config rv3028_eeprom_config_##node = {                   \
+		.i2c = rv3028_config_##rtc_inst.i2c,                                               \
+		.addr = DT_REG_ADDR(node),                                                         \
+		.size = DT_REG_SIZE(node),                                                         \
+	};                                                                                         \
+	DEVICE_DT_DEFINE(node, NULL, NULL, &rv3028_data_##rtc_inst, &rv3028_eeprom_config_##node,  \
+			 POST_KERNEL, CONFIG_RTC_INIT_PRIORITY, &rv3028_eeprom_driver_api);
+#else
+#define RV3028_EEPROM_INIT(node, rtc_inst)
+#endif /* CONFIG_RTC_RV3028_EEPROM */
```

Comment:
You can decouple the child devices from the parent more efficiently by moving most of the code of the child devices to their own source files, including their own `DT_DRV_INST` and `DT_INST_FOREACH_STATUS_OKAY()`. Then, use `DEVICE_DT_GET(DT_INST_PARENT())` to get a pointer to the parent device, and access the shared data you need that way :)

This way, you only need to share the data struct of the parent `struct rv3028_data` and your shared functions in the header file, and keep everything else static in their own source files :)

Then, moving the source files for the child devices to their respective subsystems (eeprom and retmen) becomes trivial as well :)

Reply:
You are one of the first to do so, so absolutely understandable :)